# Задание 1. Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [1]:
import pandas as pd
import numpy as np

In [2]:
columns = [
    'surgery',
    'age',
    'hospital number',
    'rectal temperature',
    'pulse',
    'respiratory rate',
    'temperature of extremities',
    'peripheral pulse',
    'mucous membranes',
    'capillary refill time',
    'pain',
    'peristalsis',
    'abdominal distension',
    'nasogastric tube',
    'nasogastric reflux',
    'nasogastric reflux PH',
    'rectal examination - feces',
    'abdomen',
    'packed cell volume',
    'total protein',
    'abdominocentesis appearance',
    'abdomcentesis total protein',
    'outcome',
    'surgical lesion',
    'type of lesion(25)',
    'type of lesion(26)',
    'type of lesion(27)',
    'cp_data',
    ]    
horse_data = pd.read_csv('horse_data.csv', names=columns, na_values='?') 
# анализ инфо и уникальных значений по столбцам показал, что пропуски в данных заполнены "?", поэтому заменяем "?" на NaN

horse_data.head()

surgery  age  hospital number  rectal temperature  pulse  respiratory rate  \
0      2.0    1           530101                38.5   66.0              28.0   
1      1.0    1           534817                39.2   88.0              20.0   
2      2.0    1           530334                38.3   40.0              24.0   
3      1.0    9          5290409                39.1  164.0              84.0   
4      2.0    1           530255                37.3  104.0              35.0   

   temperature of extremities  peripheral pulse  mucous membranes  \
0                         3.0               3.0               NaN   
1                         NaN               NaN               4.0   
2                         1.0               1.0               3.0   
3                         4.0               1.0               6.0   
4                         NaN               NaN               6.0   

   capillary refill time  ...  packed cell volume  total protein  \
0                    2.0  ...                45.0            8.4   
1                    1.0  ...                50.0           85.0   
2                    1.0  ...                33.0            6.7   
3                    2.0  ...                48.0            7.2   
4                    2.0  ...                74.0            7.4   

   abdominocentesis appearance  abdomcentesis total protein  outcome  \
0                          NaN                          NaN      2.0   
1                          2.0                          2.0      3.0   
2                          NaN                          NaN      1.0   
3                          3.0                          5.3      2.0   
4                          NaN                          NaN      2.0   

   surgical lesion  type of lesion(25)  type of lesion(26)  \
0                2               11300                   0   
1                2                2208                   0   
2                2                   0                   0   
3                1                2208                   0   
4                2                4300                   0   

   type of lesion(27)  cp_data  
0                   0        2  
1                   0        2  
2                   0        1  
3                   0        1  
4                   0        2  

[5 rows x 28 columns]

In [3]:
horse_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   surgery                      299 non-null    float64
 1   age                          300 non-null    int64  
 2   hospital number              300 non-null    int64  
 3   rectal temperature           240 non-null    float64
 4   pulse                        276 non-null    float64
 5   respiratory rate             242 non-null    float64
 6   temperature of extremities   244 non-null    float64
 7   peripheral pulse             231 non-null    float64
 8   mucous membranes             253 non-null    float64
 9   capillary refill time        268 non-null    float64
 10  pain                         245 non-null    float64
 11  peristalsis                  256 non-null    float64
 12  abdominal distension         244 non-null    float64
 13  nasogastric tube    

In [4]:
len(horse_data['hospital number'].unique())

284

In [5]:
young = horse_data.query('age == 9').groupby('surgery').count()
young.head()

age  hospital number  rectal temperature  pulse  respiratory rate  \
surgery                                                                      
1.0       18               18                  15     16                15   
2.0        6                6                   5      6                 5   

         temperature of extremities  peripheral pulse  mucous membranes  \
surgery                                                                   
1.0                              13                12                13   
2.0                               3                 2                 3   

         capillary refill time  pain  ...  packed cell volume  total protein  \
surgery                               ...                                      
1.0                         14    14  ...                  17             16   
2.0                          4     3  ...                   6              6   

         abdominocentesis appearance  abdomcentesis total protein  outcome  \
surgery                                                                      
1.0                               11                            6       18   
2.0                                2                            2        6   

         surgical lesion  type of lesion(25)  type of lesion(26)  \
surgery                                                            
1.0                   18                  18                  18   
2.0                    6                   6                   6   

         type of lesion(27)  cp_data  
surgery                               
1.0                      18       18  
2.0                       6        6  

[2 rows x 27 columns]

In [6]:
abd = horse_data.groupby('abdominal distension').count()
abd.head()

surgery  age  hospital number  rectal temperature  \
abdominal distension                                                      
1.0                        75   76               76                  69   
2.0                        65   65               65                  56   
3.0                        65   65               65                  52   
4.0                        38   38               38                  23   

                      pulse  respiratory rate  temperature of extremities  \
abdominal distension                                                        
1.0                      75                62                          70   
2.0                      59                54                          57   
3.0                      60                57                          62   
4.0                      35                25                          36   

                      peripheral pulse  mucous membranes  \
abdominal distension                                       
1.0                                 68                73   
2.0                                 56                61   
3.0                                 56                62   
4.0                                 33                34   

                      capillary refill time  ...  packed cell volume  \
abdominal distension                         ...                       
1.0                                      75  ...                  74   
2.0                                      64  ...                  63   
3.0                                      61  ...                  61   
4.0                                      36  ...                  31   

                      total protein  abdominocentesis appearance  \
abdominal distension                                               
1.0                              75                           35   
2.0                              61                           36   
3.0                              62                           36   
4.0                              28                           14   

                      abdomcentesis total protein  outcome  surgical lesion  \
abdominal distension                                                          
1.0                                            28       75               76   
2.0                                            26       65               65   
3.0                                            24       65               65   
4.0                                             8       38               38   

                      type of lesion(25)  type of lesion(26)  \
abdominal distension                                           
1.0                                   76                  76   
2.0                                   65                  65   
3.0                                   65                  65   
4.0                                   38                  38   

                      type of lesion(27)  cp_data  
abdominal distension                               
1.0                                   76       76  
2.0                                   65       65  
3.0                                   65       65  
4.0                                   38       38  

[4 rows x 27 columns]

In [7]:
outcome = horse_data.groupby(['outcome', 'surgery']).count()
outcome.head()

age  hospital number  rectal temperature  pulse  \
outcome surgery                                                    
1.0     1.0       95               95                  80     88   
        2.0       83               83                  72     78   
2.0     1.0       58               58                  39     49   
        2.0       19               19                  14     17   
3.0     1.0       27               27                  21     27   

                 respiratory rate  temperature of extremities  \
outcome surgery                                                 
1.0     1.0                    76                          75   
        2.0                    71                          71   
2.0     1.0                    45                          47   
        2.0                    13                          17   
3.0     1.0                    23                          18   

                 peripheral pulse  mucous membranes  capillary refill time  \
outcome surgery                                                              
1.0     1.0                    75                78                     84   
        2.0                    64                72                     75   
2.0     1.0                    42                49                     50   
        2.0                    17                17                     17   
3.0     1.0                    17                21                     24   

                 pain  ...  abdomen  packed cell volume  total protein  \
outcome surgery        ...                                               
1.0     1.0        75  ...       58                  88             87   
        2.0        69  ...       47                  77             78   
2.0     1.0        48  ...       31                  51             49   
        2.0        17  ...       13                  18             16   
3.0     1.0        20  ...       19                  22             22   

                 abdominocentesis appearance  abdomcentesis total protein  \
outcome surgery                                                             
1.0     1.0                               48                           32   
        2.0                               26                           20   
2.0     1.0                               31                           20   
        2.0                                5                            4   
3.0     1.0                               15                           17   

                 surgical lesion  type of lesion(25)  type of lesion(26)  \
outcome surgery                                                            
1.0     1.0                   95                  95                  95   
        2.0                   83                  83                  83   
2.0     1.0                   58                  58                  58   
        2.0                   19                  19                  19   
3.0     1.0                   27                  27                  27   

                 type of lesion(27)  cp_data  
outcome surgery                               
1.0     1.0                      95       95  
        2.0                      83       83  
2.0     1.0                      58       58  
        2.0                      19       19  
3.0     1.0                      27       27  

[5 rows x 26 columns]

In [8]:
lesion = horse_data.groupby('type of lesion(25)').count()
lesion.head()

surgery  age  hospital number  rectal temperature  pulse  \
type of lesion(25)                                                             
0                        56   56               56                  51     52   
300                       1    1                1                   1      1   
400                       5    5                5                   4      5   
1111                      1    1                1                   1      1   
1124                      1    1                1                   0      1   

                    respiratory rate  temperature of extremities  \
type of lesion(25)                                                 
0                                 48                          47   
300                                1                           1   
400                                5                           5   
1111                               0                           1   
1124                               1                           1   

                    peripheral pulse  mucous membranes  capillary refill time  \
type of lesion(25)                                                              
0                                 43                48                     49   
300                                1                 1                      1   
400                                5                 5                      5   
1111                               1                 1                      1   
1124                               1                 1                      1   

                    ...  abdomen  packed cell volume  total protein  \
type of lesion(25)  ...                                               
0                   ...       27                  52             52   
300                 ...        1                   1              1   
400                 ...        4                   5              5   
1111                ...        1                   0              0   
1124                ...        1                   1              1   

                    abdominocentesis appearance  abdomcentesis total protein  \
type of lesion(25)                                                             
0                                            17                           13   
300                                           1                            1   
400                                           2                            1   
1111                                          0                            0   
1124                                          0                            0   

                    outcome  surgical lesion  type of lesion(26)  \
type of lesion(25)                                                 
0                        56               56                  56   
300                       1                1                   1   
400                       5                5                   5   
1111                      1                1                   1   
1124                      1                1                   1   

                    type of lesion(27)  cp_data  
type of lesion(25)                               
0                                   56       56  
300                                  1        1  
400                                  5        5  
1111                                 1        1  
1124                                 1        1  

[5 rows x 27 columns]

In [9]:
req_data = horse_data.loc[:, ['surgery','age','rectal temperature','pulse','abdominal distension', 'abdomcentesis total protein', 'outcome', 'type of lesion(25)']]
req_data.head()

surgery  age  rectal temperature  pulse  abdominal distension  \
0      2.0    1                38.5   66.0                   4.0   
1      1.0    1                39.2   88.0                   2.0   
2      2.0    1                38.3   40.0                   1.0   
3      1.0    9                39.1  164.0                   4.0   
4      2.0    1                37.3  104.0                   NaN   

   abdomcentesis total protein  outcome  type of lesion(25)  
0                          NaN      2.0               11300  
1                          2.0      3.0                2208  
2                          NaN      1.0                   0  
3                          5.3      2.0                2208  
4                          NaN      2.0                4300

In [10]:
# рассчитаем базовые метрики (при этом следует учитывать, что ввиду того, что большая часть исходных данных имеет тип категориальных порядковых, базовые метрики не показательны)
req_data.describe()

surgery         age  rectal temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       abdominal distension  abdomcentesis total protein     outcome  \
count            244.000000                   102.000000  299.000000   
mean               2.266393                     3.019608    1.551839   
std                1.065131                     1.968567    0.737187   
min                1.000000                     0.100000    1.000000   
25%                1.000000                     2.000000    1.000000   
50%                2.000000                     2.250000    1.000000   
75%                3.000000                     3.900000    2.000000   
max                4.000000                    10.100000    3.000000   

       type of lesion(25)  
count          300.000000  
mean          3657.880000  
std           5399.513513  
min              0.000000  
25%           2111.750000  
50%           2673.500000  
75%           3209.000000  
max          41110.000000

In [11]:
# рассчитаем моду
def moda_calculate ():
    surgery_counts = {}
    for x in req_data.surgery.round():
        if x not in surgery_counts:
            surgery_counts[x] = 1
        else:
            surgery_counts[x] += 1
            
    max_x = 0
    mode_surgery = None
    for k, v in surgery_counts.items():
        if max_x < v:
            max_x = v
            mode_surgery = k
    print('Значение моды для surgery:', mode_surgery, 'количество встречаемости:', max_x)
    
    age_counts = {}
    for x in req_data.age.round():
        if x not in age_counts:
            age_counts[x] = 1
        else:
            age_counts[x] += 1
            
    max_x = 0
    mode_age = None
    for k, v in age_counts.items():
        if max_x < v:
            max_x = v
            mode_age = k
    print ('Значение моды для age:', mode_age, 'количество встречаемости:', max_x)
    
    temperature_counts = {}
    for x in req_data['rectal temperature'].round():
        if x not in temperature_counts:
            temperature_counts[x] = 1
        else:
            temperature_counts[x] += 1
            
    max_x = 0
    mode_temperature = None
    for k, v in temperature_counts.items():
        if max_x < v:
            max_x = v
            mode_temperature = k
    print('Значение моды для rectal temperature:', mode_temperature, 'количество встречаемости:', max_x)
    
    pulse_counts = {}
    for x in req_data.pulse.round():
        if x not in pulse_counts:
            pulse_counts[x] = 1
        else:
            pulse_counts[x] += 1
            
    max_x = 0
    mode_pulse = None
    for k, v in pulse_counts.items():
        if max_x < v:
            max_x = v
            mode_pulse = k
    print('Значение моды для pulse:', mode_pulse, 'количество встречаемости:', max_x)
    
    abdominal_distension_counts = {}
    for x in req_data['abdominal distension'].round():
        if x not in abdominal_distension_counts:
            abdominal_distension_counts[x] = 1
        else:
            abdominal_distension_counts[x] += 1
            
    max_x = 0
    mode_abdominal_distension = None
    for k, v in abdominal_distension_counts.items():
        if max_x < v:
            max_x = v
            mode_abdominal_distension = k
    print('Значение моды для abdominal distension:', mode_abdominal_distension, 'количество встречаемости:', max_x)
    
    abdomcentesis_total_protein_counts = {}
    for x in req_data['abdomcentesis total protein'].round():
        if x not in abdomcentesis_total_protein_counts:
            abdomcentesis_total_protein_counts[x] = 1
        else:
            abdomcentesis_total_protein_counts[x] += 1
    
    max_x = 0
    mode_abdomcentesis_total_protein = None
    for k, v in abdomcentesis_total_protein_counts.items():
        if max_x < v:
            max_x = v
            mode_abdomcentesis_total_protein = k
    print('Значение моды для abdomcentesis total protein:', mode_abdomcentesis_total_protein, 'количество встречаемости:', max_x)
    
    outcome_counts = {}
    for x in req_data.outcome.round():
        if x not in outcome_counts:
            outcome_counts[x] = 1
        else:
            outcome_counts[x] += 1
    
    max_x = 0
    mode_outcome = None
    for k, v in outcome_counts.items():
        if max_x < v:
            max_x = v
            mode_outcome = k
    print('Значение моды для outcome:', mode_outcome, 'количество встречаемости:', max_x)
    
    type_of_lesion_counts = {}
    for x in req_data['type of lesion(25)'].round():
        if x not in type_of_lesion_counts:
            type_of_lesion_counts[x] = 1
        else:
            type_of_lesion_counts[x] += 1
    
    max_x = 0
    mode_type_of_lesion = None
    for k, v in type_of_lesion_counts.items():
        if max_x < v:
            max_x = v
            mode_type_of_lesion = k
    print('Значение моды для type of lesion(25):', mode_type_of_lesion, 'количество встречаемости:', max_x)
    
moda_calculate()

Значение моды для surgery: 1.0 количество встречаемости: 180
Значение моды для age: 1 количество встречаемости: 276
Значение моды для rectal temperature: 38.0 количество встречаемости: 152
Значение моды для pulse: 48.0 количество встречаемости: 28
Значение моды для abdominal distension: 1.0 количество встречаемости: 76
Значение моды для abdomcentesis total protein: 2.0 количество встречаемости: 34
Значение моды для outcome: 1.0 количество встречаемости: 178
Значение моды для type of lesion(25): 0 количество встречаемости: 56


In [12]:
print('Значение моды для surgery:', req_data.surgery.mode()[0])
print('Значение моды для age:', req_data.age.mode()[0])
print('Значение моды для rectal temperature:', req_data['rectal temperature'].mode()[0])
print('Значение моды для pulse:', req_data.pulse.mode()[0])
print('Значение моды для abdominal distension:', req_data['abdominal distension'].mode()[0])
print('Значение моды для abdomcentesis total protein:', req_data['abdomcentesis total protein'].mode()[0])
print('Значение моды для outcome:', req_data.outcome.mode()[0])
print('Значение моды для type of lesion(25):', req_data['type of lesion(25)'].mode()[0])

Значение моды для surgery: 1.0
Значение моды для age: 1
Значение моды для rectal temperature: 38.0
Значение моды для pulse: 48.0
Значение моды для abdominal distension: 1.0
Значение моды для abdomcentesis total protein: 2.0
Значение моды для outcome: 1.0
Значение моды для type of lesion(25): 0


In [13]:
# рассчитаем дисперсию для rectal temperature, pulse, abdomcentesis total protein

print('rectal temperature')
print(np.var(req_data['rectal temperature'], ddof=1))
print(req_data['rectal temperature'].var())

print('pulse')
print(np.var(req_data.pulse, ddof=1))
print(req_data.pulse.var())

print('abdomcentesis total protein')
print(np.var(req_data['abdomcentesis total protein'], ddof=1))
print(req_data['abdomcentesis total protein'].var())

rectal temperature
0.5362466875871686
0.5362466875871686
pulse
819.7087747035575
819.7087747035575
abdomcentesis total protein
3.8752552902349024
3.8752552902349024


In [14]:
# межквартильный размах

Q1 = req_data['rectal temperature'].quantile(0.25)
Q3 = req_data['rectal temperature'].quantile(0.75)
IQR = Q3 - Q1
print('межквартильный размах для rectal temperature:', IQR)

Q1 = req_data.pulse.quantile(0.25)
Q3 = req_data.pulse.quantile(0.75)
IQR = Q3 - Q1
print('межквартильный размах для pulse', IQR)

Q1 = req_data['abdomcentesis total protein'].quantile(0.25)
Q3 = req_data['abdomcentesis total protein'].quantile(0.75)
IQR = Q3 - Q1
print('межквартильный размах для abdomcentesis total protein', IQR)

межквартильный размах для rectal temperature: 0.7000000000000028
межквартильный размах для pulse 40.0
межквартильный размах для abdomcentesis total protein 1.9


# Краткие выводы:
Представлена информация по 300 случаям, при этом исходя из уникальных номеров (hospital number) в исследовании участвовало 284 лошади. 
Большая часть данных имеет категориальный тип и ранжируется по группам.
Для краткого анализа были выбраны следующие столбцы: категориальные (surgery, age, abdominal distension, outcome, surgical lesion, type of lesion(25)) и количественные непрерывные (rectal temperature, pulse, abdomcentesis total protein).

Из 300 рассмотренных случаев в 180 случаях имела место операция, в 119 - лечение проходило без операции (потери минимальны, данные представлены по 299 случаям из 300). 

В данных есть описка по возрасту, согласно заданных параметров выделяются 2 категории: 1 - Adult и 2 - Young, но в представленном исследовании встречаются значения: 1 и 9. При анализе исходим из предположения, что значение 9 было ошибочно проставлено вместо значения 2.
92% (276) случаев относились к лошадям категории Adult, по 174 (63%) из них была проведена оперирация, по 101 (36,6%) - лечение без операции. Молодых лошадей (в возрасте <6 месяцев) в исследовании не много - 24 (причина такой динамики не очевидна, возможно основная целевая группа - это взрослые лошади, или маленькие жеребята меньше подвержены болезням, или по данной категории исследовались преимущественно случаи с операциями - 18 из 24 жеребят были прооперированы). 

Данные по температуре (rectal temperature) представлены только по 240 случаям (т.е. потери составляют 20%, оцениваются как существенные). Нормальной температурой считается температура = 37,8 градусов по Цельсию, в большинстве зафиксированных случаев температура была близка к нормальной(слегка повышенная), о чем свидетельствуют близкие значения меры центральной тенденции: среднее арифметическое = 38,17, медиана = 38,20, мода = 38,0 (для 152 случаев из 240). Также отметим, что болезнь скорее сопровождается повышенной температурой, понижение температуры ниже нормальной (37,8) зафиксировано менее чем в 25% случаев. Среднеквадратическое отклонение оценивается как умеренное (0,73 градуса), однако размах между min и max значениями существенный (5,4 градусов) и требует анализа на выбросы.

Данные по пульсу получены по 276 случаям (т.е. потери составляют 8%, предварительно оцениваются как умеренные), нормальным для лошадей в категории Adult признается пульс в пределах 30-40 ударов, в большинстве зафиксированных случаев пульс при болезни существенно повышен, о чем свидетельствуют все рассчитанные показатели. Следует отметить довольно существенный разброс мер центральной тенденции: среднее арифметическое = 72, медианное = 64, мода = 48 (но только для 28 случаев).

Одним из важных параметров при анализе заболеваемости считается показатель вздутия живота (abdominal distention), он был оценен для 244 случаев из 300, в остальных 56 случаях данных нет, потери существенные. Несмотря на его важность, довольно часто данный признак отсутствовал, что подтверждает значение моды = 1 - none в 76 случаях, еще в 65 случаях зафиксировано слабое вздутие (2 - slight) и только в 38 случаях зафиксировано существенное вздутие (4 - severe), свидетельствующее о том, что скорее всего нужна операция.

Показатель уровня протеина (abdomcentesis total protein), позволяющий определить вероятность нарушений кишечника зафиксирован только в 102 случаях, потери крайне высокие, причины требуют доп анализа. В связи с очень большим разбросом в данных (10 единиц между max и min) среднеарифметическое, а также медианное значения представляются нерепрезентативными, в связи с этим наиболее показательным представляется мода, которая равна 2.0 для 34 случаев из 102.

Результат лечения / операции (outcome) зафиксирован для 299 случаев, 59,5% случаев (178) завершились благополучно (лошадь выжила), в т.ч. для 95 прооперированных лошадей. В 121 случае болезнь завершилась смертью или усыплением, смертность крайне высокая. Из данных 121 случаев операция имела место быть в 85 и, очевидно, не дала положительных результатов. Предварительно можно заключить, что операция помогает только в некоторых случаях (нужна доп оценка всех показателей в комплексе) или на ранних стадиях.

Диагноз (type of lesion(25)) не определен в 56 случаях из 300 (18,7% всех случаев), диагностика оставляет желать лучшего. Основными типами заболеваний являются ущемление тонкой кишки - 51 случай, поражение толстого кишечника - 40 случаев и ущемление толстого кишечника - 30 случаев. 


# Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

# Гипотезы и работа с выбросами
Работу с выбросами в данном случае будем проводить для показателей rectal temperature, pulse, abdomcentesis total protein. По категориальным показателям все значения (кроме age где мы предполагаем, что 2 заменена 9 и принимаем это как есть) соответствуют предусмотренными классификаторам или отсутствуют.

Меры центральной тенденции по показателю rectal temperature, свидетельствуют о высокой устойчивости среднего значения: среднеарифметическое, медиана и мода максимально близки. Условные колебания в пределах от 37 до 39 градусов (диапазон выбран иходя из средних нормальных значений температуры для взрослой лошади и жеребенка, у жеребенка температура, как правило, выше) по сравнению с нормальной температурой не привели к значимому изменению мер центральной тенденции. Межквартальный размах свидетельствует о том, что 226 из 240 заполненных значений попадают в заданный диапазон, т.е. только 14 выходят за рамки, при этом min и max значения 35,4 и 40,8 соответственно, выглядят достаточно правдоподобными. В связи с чем 14 значений предлагается оставить такими какие они есть, по пропущенным значениям заполнить медианой в зависимости от возраста лошади (38.1 для взрослой лошади и 38.6 для жеребенка).

Как правило у лошади низкий пульс (по разным источникам от 28 до 44 ударов в минуту), что объясняется высокой активностью блуждающего нерва. Однако, при выделении эпинефрина (адреналина), в том числе по причине стресса, боли, кровопотери, который пересиливает тонус блуждающего нерва, пульс может быстро подняться до 220-240 уд/мин. Также необходимо учитывать существенные различия в среднем пульсе взрослой лошади (28-44 уд/мин) и жеребенка (90-120). С учетом вышеизложенного, выбросы считаются допустимыми, пропущенные значения предлагаю заполнить медианой с учетом возраста лошади и сложности заболевания (операция / лечение).

Относительно abdomcentesis total protein в исходных данных не указаны нормативные значения данного показателя, кроме примечания, что чем выше значение, тем вероятнее наличие проблем с желудком. У нас имеются данные всего по 102 случаям, при этом их диапазон существенно разниться от 0,1 до 10,1 gms/dL. Никакой логической зависимости между 7 другими показателями и уровнем abdomcentesis total protein не выявлено. Т.о. заполнение в данном случае с учетом группировки или средними значения кажется мне не обоснованным. При анализе я бы предпочла отказаться от данного показателя с учетом недостаточности данных для его трактовки.

In [15]:
# средняя температура
print(req_data['rectal temperature'].mean())
print(req_data[(req_data['rectal temperature'] > 37) & (req_data['rectal temperature'] < 39)]['rectal temperature'].mean())

38.16791666666669
38.07450000000001


In [16]:
# медиана
print(req_data['rectal temperature'].median())
print(req_data[(req_data['rectal temperature'] > 37) & (req_data['rectal temperature'] < 39)]['rectal temperature'].median())

38.2
38.1


In [17]:
# мода
print(req_data['rectal temperature'].mode()[0])
print(req_data[(req_data['rectal temperature'] > 37) & (req_data['rectal temperature'] < 39)]['rectal temperature'].round().mode()[0])

38.0
38.0


In [18]:
# межквартальный размах по температуре
q1 = req_data['rectal temperature'].quantile(0.25)
q3 = req_data['rectal temperature'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = req_data[req_data['rectal temperature'].between(lower_bound, upper_bound, inclusive=True)].sort_values('age')
remove_outliers

surgery  age  rectal temperature  pulse  abdominal distension  \
0        2.0    1                38.5   66.0                   4.0   
189      1.0    1                38.3   60.0                   1.0   
190      1.0    1                37.1   40.0                   1.0   
192      1.0    1                38.2   48.0                   3.0   
194      2.0    1                37.9   88.0                   1.0   
..       ...  ...                 ...    ...                   ...   
255      1.0    9                38.8  184.0                   3.0   
252      1.0    9                38.8  124.0                   4.0   
39       1.0    9                39.2  146.0                   NaN   
135      1.0    9                38.1  136.0                   3.0   
212      1.0    9                38.9  120.0                   3.0   

     abdomcentesis total protein  outcome  type of lesion(25)  
0                            NaN      2.0               11300  
189                          3.0      1.0               31110  
190                          NaN      1.0                3133  
192                          2.0      1.0                5111  
194                          NaN      1.0                   0  
..                           ...      ...                 ...  
255                          NaN      2.0                7111  
252                          NaN      2.0                2208  
39                           NaN      2.0                2113  
135                          2.9      2.0                2205  
212                          NaN      1.0                2124  

[226 rows x 8 columns]

In [19]:
pd.concat([req_data, remove_outliers]).drop_duplicates(keep=False)

surgery  age  rectal temperature  pulse  abdominal distension  \
7        1.0    1                 NaN   60.0                   2.0   
8        2.0    1                 NaN   80.0                   4.0   
16       1.0    9                 NaN  128.0                   3.0   
20       1.0    1                39.9   72.0                   4.0   
28       1.0    1                 NaN    NaN                   NaN   
..       ...  ...                 ...    ...                   ...   
282      1.0    1                 NaN   70.0                   2.0   
288      1.0    1                 NaN    NaN                   NaN   
293      1.0    1                 NaN   78.0                   NaN   
295      1.0    1                 NaN  120.0                   NaN   
298      1.0    1                36.5  100.0                   3.0   

     abdomcentesis total protein  outcome  type of lesion(25)  
7                            NaN      2.0                2208  
8                            NaN      3.0                3205  
16                           4.7      2.0                1400  
20                           NaN      1.0                2111  
28                           NaN      2.0                4300  
..                           ...      ...                 ...  
282                          NaN      2.0                1400  
288                          NaN      1.0                2124  
293                          2.0      3.0                2209  
295                          NaN      3.0                3205  
298                          3.4      1.0                2208  

[69 rows x 8 columns]

In [20]:
# медиана по температуре в зависимости от возраста лошади
req_data.groupby('age').agg({'rectal temperature': 'median'})['rectal temperature'].head()

age
1    38.1
9    38.6
Name: rectal temperature, dtype: float64

In [21]:
# межквартальный размах по пульсу
q1 = req_data['pulse'].quantile(0.25)
q3 = req_data['pulse'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers_p = req_data[req_data['pulse'].between(lower_bound, upper_bound, inclusive=True)].sort_values('surgery')
remove_outliers_p

surgery  age  rectal temperature  pulse  abdominal distension  \
152      1.0    1                 NaN  100.0                   4.0   
172      1.0    1                38.5  129.0                   3.0   
173      1.0    1                 NaN  104.0                   3.0   
176      1.0    1                38.2   60.0                   1.0   
177      1.0    1                 NaN   68.0                   NaN   
..       ...  ...                 ...    ...                   ...   
66       2.0    1                37.9   72.0                   1.0   
67       2.0    1                38.4   54.0                   1.0   
236      2.0    1                38.2   42.0                   2.0   
0        2.0    1                38.5   66.0                   4.0   
132      NaN    1                38.0   48.0                   1.0   

     abdomcentesis total protein  outcome  type of lesion(25)  
152                          2.0      1.0                2209  
172                          2.0      1.0                2206  
173                          3.4      2.0                2207  
176                          NaN      1.0                6112  
177                          2.8      2.0                4205  
..                           ...      ...                 ...  
66                           2.0      1.0                2322  
67                           NaN      1.0                3111  
236                          NaN      1.0                3111  
0                            NaN      2.0               11300  
132                          1.0      NaN                3111  

[271 rows x 8 columns]

In [22]:
pd.concat([req_data, remove_outliers_p]).drop_duplicates(keep=False)

surgery  age  rectal temperature  pulse  abdominal distension  \
3        1.0    9                39.1  164.0                   4.0   
28       1.0    1                 NaN    NaN                   NaN   
41       2.0    9                39.0  150.0                   NaN   
52       2.0    1                 NaN    NaN                   1.0   
55       1.0    9                38.6  160.0                   4.0   
56       1.0    1                 NaN    NaN                   NaN   
58       1.0    1                 NaN    NaN                   3.0   
74       1.0    9                 NaN    NaN                   NaN   
83       1.0    1                38.0    NaN                   4.0   
93       2.0    1                 NaN    NaN                   3.0   
117      2.0    1                39.5    NaN                   3.0   
126      1.0    1                38.7    NaN                   2.0   
151      1.0    1                39.3    NaN                   2.0   
160      1.0    1                 NaN    NaN                   NaN   
174      2.0    1                 NaN    NaN                   4.0   
175      1.0    1                 NaN    NaN                   NaN   
182      1.0    1                38.0    NaN                   4.0   
204      1.0    1                39.2    NaN                   2.0   
207      1.0    1                 NaN    NaN                   NaN   
216      2.0    1                 NaN    NaN                   NaN   
227      1.0    9                 NaN    NaN                   NaN   
255      1.0    9                38.8  184.0                   3.0   
275      1.0    9                38.8  150.0                   2.0   
288      1.0    1                 NaN    NaN                   NaN   

     abdomcentesis total protein  outcome  type of lesion(25)  
3                            5.3      2.0                2208  
28                           NaN      2.0                4300  
41                           0.1      1.0                9400  
52                           NaN      1.0                3111  
55                           NaN      2.0                7111  
56                           NaN      1.0                3112  
58                           NaN      2.0                3205  
74                           NaN      2.0               11124  
83                           NaN      2.0                3205  
93                           NaN      2.0                3205  
117                          NaN      3.0                4205  
126                          NaN      1.0               31110  
151                          4.3      2.0                2207  
160                          1.0      1.0                4205  
174                          NaN      2.0                3400  
175                          NaN      1.0                2208  
182                          NaN      2.0                2205  
204                          3.0      1.0                3115  
207                          NaN      2.0                2208  
216                          NaN      1.0                   0  
227                          NaN      1.0                9000  
255                          NaN      2.0                7111  
275                          NaN      2.0                4207  
288                          NaN      1.0                2124

In [23]:
# медиана по пульсу в зависимости от сложности болезни (операция / лечение) для лошади
req_data.groupby(['age', 'surgery']).agg({'pulse': 'median'})['pulse'].head()

age  surgery
1    1.0         66.0
     2.0         54.0
9    1.0        129.0
     2.0         91.0
Name: pulse, dtype: float64

In [24]:
# медиана по пульсу в зависимости от сложности болезни (операция / лечение) для жеребенка
req_data.groupby(['age', 'surgery']).agg({'pulse': 'mean'})['pulse'].head()

age  surgery
1    1.0         70.398649
     2.0         63.495238
9    1.0        131.875000
     2.0        100.666667
Name: pulse, dtype: float64

In [25]:
# группировка показателей по содержанию белка в зависимости от выбранного метода лечения и конечного результата
req_data.groupby(['surgery', 'outcome']).agg({'abdomcentesis total protein': ['median', 'mean', 'min', 'max']})['abdomcentesis total protein']

median      mean  min   max
surgery outcome                             
1.0     1.0        2.00  2.968750  1.0   7.0
        2.0        3.40  4.075000  1.4  10.1
        3.0        2.00  1.764706  1.0   2.0
2.0     1.0        2.15  2.655000  0.1   8.0
        2.0        4.30  4.550000  3.6   6.0
        3.0        3.60  3.650000  1.0   7.4

In [26]:
req_data.groupby(['surgery', 'abdominal distension']).count()

age  rectal temperature  pulse  \
surgery abdominal distension                                   
1.0     1.0                    34                  32     34   
        2.0                    31                  26     28   
        3.0                    50                  40     47   
        4.0                    27                  17     25   
2.0     1.0                    41                  36     40   
        2.0                    34                  30     31   
        3.0                    15                  12     13   
        4.0                    11                   6     10   

                              abdomcentesis total protein  outcome  \
surgery abdominal distension                                         
1.0     1.0                                            16       34   
        2.0                                            16       31   
        3.0                                            18       50   
        4.0                                             7       27   
2.0     1.0                                            11       41   
        2.0                                            10       34   
        3.0                                             6       15   
        4.0                                             1       11   

                              type of lesion(25)  
surgery abdominal distension                      
1.0     1.0                                   34  
        2.0                                   31  
        3.0                                   50  
        4.0                                   27  
2.0     1.0                                   41  
        2.0                                   34  
        3.0                                   15  
        4.0                                   11

# Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [27]:
for col in req_data.columns:
    pct_missing = req_data[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery - 0.3%
age - 0.0%
rectal temperature - 20.0%
pulse - 8.0%
abdominal distension - 18.7%
abdomcentesis total protein - 66.0%
outcome - 0.3%
type of lesion(25) - 0.0%


# Работа с пропусками:
по 2 параметрам пропуски отсутствуют (age и type of lesion(25)).

по 2 параметрам (surgery и outcome) допущен всего 1 пропуск по одному и тому же случаю (строке) - ввиду незначительного удельного веса относительно общей выборки и с учетом большого разброса значений в type of lesion(25), предлагаю данный случай удалить.

по показателю rectal temperature, ввиду незначительного расхождения мер центральной тенденции и среднеквадратичного отклонения, предлагаю заменить пропуски медианой с учетом группировки по столбцу age.

по показателю pulse, заполнить пропуски медианой с учетом группировки по столбцам age и surgery (обоснование см. в задании 2).

по показателю abdominal distension - отсутствующие данные предлагаю заполнить константой ("n/a") ввиду отсутствия логической взаимосвязи и динамики в сравнении с другими показателями.

по показателю abdomcentesis total protein - предлагаю удалить столбец с учетом комментариев в задании №2.

In [28]:
new_df = horse_data.loc[:, ['surgery','age','rectal temperature','pulse','abdominal distension', 'abdomcentesis total protein', 'outcome', 'type of lesion(25)']]
new_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   surgery                      299 non-null    float64
 1   age                          300 non-null    int64  
 2   rectal temperature           240 non-null    float64
 3   pulse                        276 non-null    float64
 4   abdominal distension         244 non-null    float64
 5   abdomcentesis total protein  102 non-null    float64
 6   outcome                      299 non-null    float64
 7   type of lesion(25)           300 non-null    int64  
dtypes: float64(6), int64(2)
memory usage: 18.9 KB


In [29]:
# удалим строку с пропущенным значением surgery и outcome
df = new_df.drop(index=[132])
#df.info()

In [30]:
# удалим столбец abdomcentesis total protein
df_1 = df.drop(['abdomcentesis total protein'], axis=1)
df_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 299 entries, 0 to 299
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               299 non-null    float64
 1   age                   299 non-null    int64  
 2   rectal temperature    239 non-null    float64
 3   pulse                 275 non-null    float64
 4   abdominal distension  243 non-null    float64
 5   outcome               299 non-null    float64
 6   type of lesion(25)    299 non-null    int64  
dtypes: float64(5), int64(2)
memory usage: 18.7 KB


In [31]:
# заменим пропуски в столбце abdominal distension на n/a
df_1['abdominal distension'].fillna('n/a', inplace = True)
df_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 299 entries, 0 to 299
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               299 non-null    float64
 1   age                   299 non-null    int64  
 2   rectal temperature    239 non-null    float64
 3   pulse                 275 non-null    float64
 4   abdominal distension  299 non-null    object 
 5   outcome               299 non-null    float64
 6   type of lesion(25)    299 non-null    int64  
dtypes: float64(4), int64(2), object(1)
memory usage: 18.7+ KB


In [33]:
print(df_1.groupby('age')['rectal temperature'].median())
print(df_1.groupby('age')['rectal temperature'].mean())
print(df_1['rectal temperature'].mode()[0])

age
1    38.1
9    38.6
Name: rectal temperature, dtype: float64
age
1    38.125114
9    38.645000
Name: rectal temperature, dtype: float64
38.0


In [37]:
# заполнение rectal temperature медианой для групп в зависимости от возраста
df_1['rectal temperature'].fillna(df_1.groupby('age')['rectal temperature'].transform('median'), inplace=True)
print(df_1.groupby('age')['rectal temperature'].median())
print(df_1.groupby('age')['rectal temperature'].mean())
print(df_1['rectal temperature'].mode()[0])

age
1    38.1
9    38.6
Name: rectal temperature, dtype: float64
age
1    38.1200
9    38.6375
Name: rectal temperature, dtype: float64
38.1


In [38]:
print(df_1.groupby(['age', 'surgery']).pulse.median())
print(df_1.groupby(['age', 'surgery']).pulse.mean())
print(df_1.pulse.mode()[0])

age  surgery
1    1.0         66.0
     2.0         54.0
9    1.0        129.0
     2.0         91.0
Name: pulse, dtype: float64
age  surgery
1    1.0         70.398649
     2.0         63.495238
9    1.0        131.875000
     2.0        100.666667
Name: pulse, dtype: float64
48.0


In [39]:
df_1.pulse.fillna(df_1.groupby(['age', 'surgery']).pulse.transform('median'), inplace=True)
print(df_1.groupby(['age', 'surgery']).pulse.median())
print(df_1.groupby(['age', 'surgery']).pulse.mean())
print(df_1.pulse.mode()[0])

age  surgery
1    1.0         66.0
     2.0         54.0
9    1.0        129.0
     2.0         91.0
Name: pulse, dtype: float64
age  surgery
1    1.0         70.018519
     2.0         62.823009
9    1.0        131.555556
     2.0        100.666667
Name: pulse, dtype: float64
48.0


In [41]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 299 entries, 0 to 299
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               299 non-null    float64
 1   age                   299 non-null    int64  
 2   rectal temperature    299 non-null    float64
 3   pulse                 299 non-null    float64
 4   abdominal distension  299 non-null    object 
 5   outcome               299 non-null    float64
 6   type of lesion(25)    299 non-null    int64  
dtypes: float64(4), int64(2), object(1)
memory usage: 28.7+ KB
